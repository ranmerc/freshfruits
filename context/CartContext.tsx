import { useLocalStorage } from "usehooks-ts";
import {
  createContext,
  useContext,
  ReactNode,
  useReducer,
  Dispatch,
  useCallback,
} from "react";
import CartItem from "@/types/CartItem";

type ActionType =
  | {
      type: "ADD_TO_CART";
      payload: CartItem;
    }
  | {
      type: "REMOVE_FROM_CART";
      payload: {
        fruitId: number;
        selectedPackId: number;
      };
    }
  | {
      type: "CLEAR_CART";
    };

interface CartContextType {
  cartItems: CartItem[];
  cartDispatch: Dispatch<ActionType>;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export function useCartContext() {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }

  return context;
}

function cartReducer(state: CartItem[], action: ActionType) {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { fruitId, selectedPackId, quantity } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex !== -1) {
        const newState = [...state];
        newState[existingItemIndex].quantity += quantity;
        return newState;
      }

      return [...state, action.payload];
    }

    case "REMOVE_FROM_CART": {
      const { fruitId, selectedPackId } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex === -1) {
        return state;
      }

      const newState = [...state];
      newState[existingItemIndex].quantity -= 1;

      if (newState[existingItemIndex].quantity === 0) {
        newState.splice(existingItemIndex, 1);
      }

      return newState;
    }

    case "CLEAR_CART":
      return [];
    default:
      return state;
  }
}

export function CartProvider({ children }: { children: ReactNode }) {
  const [savedCartItems, setSavedCartItems] = useLocalStorage<CartItem[]>(
    "cart-items",
    []
  );

  const reducerLocalStorage = useCallback(
    (state: CartItem[], action: ActionType) => {
      const newState = cartReducer(state, action);

      setSavedCartItems(newState);

      return newState;
    },
    [setSavedCartItems]
  );

  const [state, dispatch] = useReducer(reducerLocalStorage, savedCartItems);

  return (
    <CartContext.Provider value={{ cartItems: state, cartDispatch: dispatch }}>
      {children}
    </CartContext.Provider>
  );
}
