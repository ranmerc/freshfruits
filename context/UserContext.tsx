import { createContext, useContext, ReactNode } from "react";
import { useLocalStorage } from "usehooks-ts";

export interface UserContextType {
  username: string;
  setUserName: (user: string) => void;
}

export const UserContext = createContext<UserContextType | undefined>(
  undefined
);

export function useUserContext() {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return context;
}

export function UserProvider({ children }: { children: ReactNode }) {
  const [username, setUserName] = useLocalStorage<string>("username", "");

  return (
    <UserContext.Provider value={{ username, setUserName }}>
      {children}
    </UserContext.Provider>
  );
}
