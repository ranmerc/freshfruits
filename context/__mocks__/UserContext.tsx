const UserContext = jest.fn();

export default UserContext;
