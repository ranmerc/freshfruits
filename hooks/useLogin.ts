import { useMutation } from "@tanstack/react-query";
import axios, { AxiosResponse } from "axios";
import { ErrorResponse, FormValues, SuccessResponse } from "@/types/LoginTypes";

const useLogin = () => {
  return useMutation({
    mutationFn: async (values: FormValues) => {
      const response = await axios.post<
        FormValues,
        AxiosResponse<SuccessResponse>
      >("/api/login", values);
      return response.data.username;
    },
  });
};

export default useLogin;
