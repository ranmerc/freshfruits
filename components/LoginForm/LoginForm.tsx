import { useUserContext } from "@/context/UserContext";
import LoginSchema from "@/schemas/LoginSchema";
import { Button, Stack, TextField, Typography } from "@mui/material";
import { Field, Form, Formik, FormikHelpers } from "formik";
import { Alert, Backdrop, CircularProgress, Snackbar } from "@mui/material";
import useLogin from "@/hooks/useLogin";
import { AccountCircle } from "@mui/icons-material";
import Link from "next/link";
import { useEffect, useState } from "react";
import { FormValues } from "@/types/LoginTypes";
import axios from "axios";

export default function LoginForm() {
  const { isLoading, isError, error, mutateAsync } = useLogin();
  const { setUserName } = useUserContext();
  const [showError, setShowError] = useState(isError);

  useEffect(() => {
    setShowError(isError);
  }, [isError]);

  const handleLogin = async (
    values: FormValues,
    actions: FormikHelpers<FormValues>
  ) => {
    const user = await mutateAsync(values);
    setUserName(user);
  };

  return (
    <>
      <Stack alignItems={"center"} width={"100%"} rowGap={2} mb={4}>
        <AccountCircle sx={{ fontSize: 60 }} color="secondary" />
        <Typography variant="h5">Login</Typography>
      </Stack>
      <Formik
        initialValues={{ email: "", password: "" }}
        validationSchema={LoginSchema}
        onSubmit={handleLogin}
      >
        {({ errors, touched }) => (
          <Form>
            <Stack
              width={{ xs: "100%", md: "40%" }}
              margin={"0 auto"}
              rowGap={3}
            >
              <Field
                as={TextField}
                label="Email Address"
                type="email"
                id="email"
                name="email"
                variant="outlined"
                error={touched.email && !!errors.email}
                helperText={touched.email && errors.email}
                required
              />
              <Field
                as={TextField}
                label="Password"
                type="password"
                id="password"
                name="password"
                variant="outlined"
                error={touched.password && !!errors.password}
                helperText={touched.password && errors.password}
                required
              />
              <Button
                type="submit"
                variant="contained"
                color="primary"
                sx={{ p: 1 }}
              >
                Login
              </Button>
              <Link href="/signup" style={{ alignSelf: "center" }}>
                <Button variant="text">Create Account</Button>
              </Link>
            </Stack>
          </Form>
        )}
      </Formik>
      <Backdrop
        sx={{ color: "#fff", zIndex: (theme) => theme.zIndex.drawer + 1 }}
        open={isLoading}
      >
        <CircularProgress color="secondary" />
      </Backdrop>
      <Snackbar
        anchorOrigin={{ vertical: "top", horizontal: "center" }}
        open={showError}
        autoHideDuration={3000}
        onClose={() => {
          setShowError(false);
        }}
        onClick={() => {
          setShowError(false);
        }}
      >
        <Alert severity="error" sx={{ width: "100%" }}>
          {axios.isAxiosError(error) && error.response?.data?.message}
        </Alert>
      </Snackbar>
    </>
  );
}
