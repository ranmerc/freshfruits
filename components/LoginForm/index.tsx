import axios from "axios";
import Link from "next/link";
import useLogin from "@/hooks/useLogin";
import FormHead from "../FormHead";
import { Person } from "@mui/icons-material";
import { Field, Form, Formik } from "formik";
import LoginSchema from "@/schemas/LoginSchema";
import { LoginFormValues } from "@/types/LoginTypes";
import { useUserContext } from "@/context/UserContext";
import { Button, Stack, TextField } from "@mui/material";
import AlertSnackbar from "../AlertSnackbar/AlertSnackbar";
import LoadingBackdrop from "../LoadingBackdrop/LoadingBackdrop";

export default function LoginForm() {
  const { isLoading, isError, error, mutateAsync } = useLogin();
  const { setUserName } = useUserContext();

  const handleLogin = async (values: LoginFormValues) => {
    const user = await mutateAsync(values);
    setUserName(user);
  };

  return (
    <>
      <FormHead SVGIcon={Person} title="Login" />
      <Formik
        initialValues={{ email: "", password: "" }}
        validationSchema={LoginSchema}
        onSubmit={handleLogin}
      >
        {({ errors, touched, isValid }) => (
          <Form>
            <Stack
              width={{ xs: "100%", md: "40%" }}
              margin={"0 auto"}
              rowGap={3}
            >
              <Field
                as={TextField}
                label="Email Address"
                type="email"
                id="email"
                name="email"
                variant="outlined"
                error={touched.email && !!errors.email}
                helperText={touched.email && errors.email}
                required
              />
              <Field
                as={TextField}
                label="Password"
                type="password"
                id="password"
                name="password"
                variant="outlined"
                error={touched.password && !!errors.password}
                helperText={touched.password && errors.password}
                required
              />
              <Button
                type="submit"
                variant="contained"
                color="primary"
                sx={{ p: 1 }}
                disabled={!isValid}
              >
                Login
              </Button>
              <Link href="/signup" style={{ alignSelf: "center" }}>
                <Button variant="text">Create Account</Button>
              </Link>
            </Stack>
          </Form>
        )}
      </Formik>
      <LoadingBackdrop open={isLoading} />
      <AlertSnackbar
        isOpen={isError}
        severity="error"
        message={axios.isAxiosError(error) && error.response?.data?.message}
      />
    </>
  );
}
