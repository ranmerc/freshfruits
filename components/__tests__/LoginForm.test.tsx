import { screen, render, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import LoginForm from "../LoginForm";
import useLogin from "@/hooks/useLogin";
import { useUserContext } from "@/context/UserContext";
import { fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

jest.mock("@/hooks/useLogin");
jest.mock("@/context/UserContext");

const mockedUseLogin = useLogin as jest.Mock;
const mockedUseUserContext = useUserContext as jest.Mock;

mockedUseLogin.mockReturnValue({
  isLoading: false,
  isError: false,
  error: null,
  mutateAsync: jest.fn(),
});

mockedUseUserContext.mockReturnValue({
  setUserName: jest.fn(),
});

describe("LoginForm component", () => {
  beforeEach(() => {
    jest.spyOn(console, "error").mockImplementation(() => {});
  });

  it("Renders form heading", () => {
    render(<LoginForm />);

    expect(screen.getByRole("heading", { name: "Login" })).toBeInTheDocument();
  });

  it("Renders input fields", () => {
    render(<LoginForm />);

    expect(
      screen.getByRole("textbox", { name: /Email Address/i })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/Password/i)).toBeInTheDocument();
  });

  it("Renders login button and register link", () => {
    render(<LoginForm />);

    expect(screen.getByRole("button", { name: "Login" })).toBeInTheDocument();

    expect(
      screen.getByRole("link", { name: "Create Account" })
    ).toHaveAttribute("href", "/signup");
  });

  it("Renders password errors", async () => {
    render(<LoginForm />);

    const passwordInput = screen.getByLabelText(/Password/i);

    expect(screen.queryByText(/Password is required/i)).not.toBeInTheDocument();

    act(() => {
      fireEvent.focus(passwordInput);
      fireEvent.blur(passwordInput);
    });

    await waitFor(() => {
      expect(screen.getByText(/Password is required/i)).toBeInTheDocument();
    });

    act(() => {
      userEvent.type(passwordInput, "test");
      fireEvent.blur(passwordInput);
    });

    await waitFor(() => {
      expect(
        screen.getByText(/Password must be at least 8 characters long/i)
      ).toBeInTheDocument();
    });

    act(() => {
      userEvent.clear(passwordInput);
      userEvent.type(passwordInput, "password");
    });

    await waitFor(() => {
      expect(
        screen.queryByText(/Password is required/i)
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText(/Password must be at least 8 characters long/i)
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText(/Password must be at most 128 characters long/i)
      ).not.toBeInTheDocument();
    });
  });

  it("Renders email errors", async () => {
    render(<LoginForm />);

    const emailInput = screen.getByRole("textbox", { name: "Email Address" });

    expect(screen.queryByText(/Email is required/i)).not.toBeInTheDocument();

    act(() => {
      fireEvent.focus(emailInput);
      fireEvent.blur(emailInput);
    });

    await waitFor(() => {
      expect(screen.getByText(/Email is required/i)).toBeInTheDocument();
    });

    act(async () => {
      userEvent.type(emailInput, "test");
      fireEvent.blur(emailInput);
    });

    await waitFor(() => {
      expect(screen.getByText(/Invalid email format/i)).toBeInTheDocument();
    });

    act(() => {
      userEvent.clear(emailInput);
      userEvent.type(emailInput, "name@email.com");
    });

    await waitFor(() => {
      expect(screen.queryByText(/Email is required/i)).not.toBeInTheDocument();
      expect(screen.queryByText(/Invalid email/i)).not.toBeInTheDocument();
    });
  });
});
