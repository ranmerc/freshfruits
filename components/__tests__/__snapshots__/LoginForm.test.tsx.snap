// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component Renders correctly 1`] = `
[
  <div
    className="MuiStack-root css-440ut8-MuiStack-root"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-9rgyfn-MuiSvgIcon-root"
      data-testid="PersonIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
      />
    </svg>
    <h5
      className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      Login
    </h5>
  </div>,
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="MuiStack-root css-8ayx1v-MuiStack-root"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email Address
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="email"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Email Address
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-upd4p9-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        Login
      </button>
      <a
        href="/signup"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "alignSelf": "center",
          }
        }
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Create Account
        </button>
      </a>
    </div>
  </form>,
  <div
    aria-hidden={true}
    className="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-1szo49z-MuiCircularProgress-root"
      role="progressbar"
      style={
        {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={{}}
        />
      </svg>
    </span>
  </div>,
]
`;
